### PR TITLE
sharedbytes: use custom bytes type to reduce memory allocation.

### PR DESCRIPTION
**Pull request by coocood**

### CODE REVIEW

**Comment by coocood** [comment]
@shenli @tiancaiamao PTAL

**Comment by c4pt0r** [comment]
How about some benchmarks?

**Comment by tiancaiamao** [comment]
LGTM

File: sharedbytes/sharedbytes_test.go [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+	}
+	x = make([]byte, 3)
+	sb.MarshalTo(x)
+	if !bytes.Equal(x, []byte("abc")) {
+		t.Fatal("marshal failed")
+	}
+	sb = nil
+	if err := sb.Unmarshal([]byte("abc")); err != nil {
+		t.Fatal(err)
+	}
```
Comment (zimulala) [inline sharedbytes/sharedbytes_test.go]: Check the `sb` value.

**Comment by coocood** [comment]
@c4pt0r 
benmark can only show that there is no memory allocation when calling Unmarshal.

**Comment by zimulala** [comment]
LGTM

**Pull request merged**